### PR TITLE
[W5.11][F09-A2]Ewald Hew

### PR DESCRIPTION
**Pull request by ewaldhew**

### CODE REVIEW

File: src/seedu/addressbook/commands/EditCommand.java [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+                    target.getTags()
+            );
+
+            addressBook.editPerson(target, replacement);
+            return new CommandResult(String.format(MESSAGE_EDIT_PERSON_SUCCESS, target, replacement));
+
+        } catch (IndexOutOfBoundsException ie) {
+            return new CommandResult(Messages.MESSAGE_INVALID_PERSON_DISPLAYED_INDEX);
+        } catch (DuplicatePersonException dpe) {
+            return new CommandResult(MESSAGE_DUPLICATE_PERSON);
```
Comment (jonathanwj) [inline src/seedu/addressbook/commands/EditCommand.java:69]: Just a question, why not put MESSAGE_DUPLICATE_PERSON string in the Messages class as well?

**Review by jonathanwj — COMMENTED** [review]
Code looks good to me.

File: src/seedu/addressbook/data/AddressBook.java [hunk 2]
```diff
@@ -83,6 +83,29 @@ public void addPerson(Person toAdd) throws DuplicatePersonException {
...
     }
 
+    /**
+     * Edits a person in the address book.
+     *
+     * @throws DuplicatePersonException if an equivalent person already exists.
+     */
+    public void editPerson(ReadOnlyPerson target, Person replacement)
+            throws DuplicatePersonException, PersonNotFoundException {
+        addPerson(replacement);
```
Comment (kokonguyen191) [inline src/seedu/addressbook/data/AddressBook.java]: Is this the wanted behavior? **AddressBook** throws **DuplicatePersonException** if it finds a person with the same name and phone number in the list. Now what if I want to change John Doe's email address from _johnd@gmail.com_ to _johnd@u.nus.edu_? Then **AddressBook** will throw **DuplicatePersonException**. Should you add first then remove or should you remove first then add?

File: src/seedu/addressbook/commands/EditCommand.java [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+import seedu.addressbook.data.exception.IllegalValueException;
+import seedu.addressbook.data.person.Address;
+import seedu.addressbook.data.person.Email;
+import seedu.addressbook.data.person.Person;
+import seedu.addressbook.data.person.Phone;
+import seedu.addressbook.data.person.ReadOnlyPerson;
+import seedu.addressbook.data.person.UniquePersonList.DuplicatePersonException;
+import seedu.addressbook.data.person.UniquePersonList.PersonNotFoundException;
+
+
```
Comment (kokonguyen191) [inline src/seedu/addressbook/commands/EditCommand.java]: Header comments

File: src/seedu/addressbook/commands/EditCommand.java [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+public class EditCommand extends Command {
```
Comment (kokonguyen191) [inline src/seedu/addressbook/commands/EditCommand.java]: Header comment

File: src/seedu/addressbook/data/AddressBook.java (hunk 2 above)
Comment (ewaldhew) [reply src/seedu/addressbook/data/AddressBook.java]: Hmm, I didn't check that duplicate considers name and phone only... intention was to disallow editing a person to be exactly same as another. I'll look into using the other compare function instead.

File: src/seedu/addressbook/commands/EditCommand.java (hunk 1 above)
Comment (ewaldhew) [reply src/seedu/addressbook/commands/EditCommand.java:69]: Not exactly the same message, should they be? hmm

Comment (kokonguyen191) [reply src/seedu/addressbook/commands/EditCommand.java:69]: If you use MESSAGE_DUPLICATE_PERSON in another **Command**, then I think putting it in **Messages** class is a good idea, else you can just put it here.

File: doc/UserGuide.md [hunk 5]
```diff
@@ -66,12 +66,29 @@ Examples:
...
+>
+> Put a `p` before the phone / email / address prefixes to mark it as `private`. `private` details can only
+> be seen using the `viewall` command.
+
+Examples:
+* `list`<br>
+  `edit 2 p/98765432 e/johnd@gmail.com a/John street, block 123, #01-01`
+  Edits the 2nd person in the address book.
+* `find Betsy`<br>
+  `edit 1 pp/1234567 e/betsycrowe@gmail.com pa/Newgate Prison t/criminal t/friend`
```
Comment (k-l-a) [inline doc/UserGuide.md:84]: Typo in `pp`?

File: src/seedu/addressbook/commands/EditCommand.java [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+    private final Phone replacementPhone;
+    private final Email replacementEmail;
+    private final Address replacementAddress;
+
+    /**
+     * Construct the edit command.
+     *
+     * @throws IllegalValueException if any of the raw values are invalid
+     */
+    public EditCommand(int targetVisibleIndex,
```
Comment (k-l-a) [inline src/seedu/addressbook/commands/EditCommand.java:40]: Are all fields mandatory? Maybe you can find a way to only change select fields.

File: src/seedu/addressbook/commands/EditCommand.java [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+        super(targetVisibleIndex);
+
+        replacementPhone = new Phone(phone, isPhonePrivate);
+        replacementEmail = new Email(email, isEmailPrivate);
+        replacementAddress = new Address(address, isAddressPrivate);
+    }
+
+    @Override
+    public CommandResult execute() {
+        try {
```
Comment (k-l-a) [inline src/seedu/addressbook/commands/EditCommand.java:53]: Try catch block should as close to the line where possible exception happens as possible.

File: test/java/seedu/addressbook/commands/EditCommandTest.java [hunk 8]
```diff
@@ -0,0 +1,287 @@
...
+import seedu.addressbook.data.person.Name;
+import seedu.addressbook.data.person.Person;
+import seedu.addressbook.data.person.Phone;
+import seedu.addressbook.data.person.ReadOnlyPerson;
+import seedu.addressbook.data.person.UniquePersonList.PersonNotFoundException;
+import seedu.addressbook.data.tag.UniqueTagList;
+import seedu.addressbook.ui.TextUi;
+import seedu.addressbook.util.TestUtil;
+
+public class EditCommandTest {
```
Comment (k-l-a) [inline test/java/seedu/addressbook/commands/EditCommandTest.java:27]: Good job creating JUnit tests for newly added method/classes.

**Review by k-l-a — COMMENTED** [review]
Good attempt! Some comments added. Please close PR after reading.

To note, please give a brief summary of what you have enhanced in your PR description. It helps understanding what enhancements you have (or intended to) made.

**Pull request closed without merging**